src/main/scala/strawman/collection/mutable/ArrayBuffer.scala

```diff
@@ -39,7 +39,7 @@ class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
         start = 0
       }
       else {
-        val newelems = new Array[AnyRef](end * 2)
+        val newelems = new Array[AnyRef](if (end == 0) 16 else (end * 2))
```

Comment:
I think this algorithm could be improved in three aspects.  First, allocating size 1 then 2 then 4 then 8 then 16 is pretty wasteful.  I'd go for at _least_ 4, maybe 8, on the first realloc.  Second, no point doing an `Array.copy` if `end==0`.  Third, you max out at (1 << 30) elements this way when you could have ((1 << 31) - 1) if you handled the last buffer increase separately.  I know you didn't write the entire code, but if we're improving this, let's do it all the way.